config-service-impl/src/main/java/org/hypertrace/config/service/store/DocumentConfigStore.java

```diff
@@ -64,68 +80,69 @@ private Datastore initDataStore(Config config) {
-      long creationTimestamp;
-      ContextSpecificConfig existingConfig = getConfig(configResource, configVersion - 1);
-      if (ConfigServiceUtils.isNull(existingConfig.getConfig())) {
-        creationTimestamp = updateTimestamp;
-      } else {
-        creationTimestamp = existingConfig.getCreationTimestamp();
-      }
-      Key key = new ConfigDocumentKey(configResource, configVersion);
-      ConfigDocument configDocument =
+    synchronized (configResourceContextLocks.getUnchecked(configResourceContext)) {
+      Optional<ConfigDocument> previousConfigDoc = getLatestVersionConfigDoc(configResourceContext);
+      Optional<ContextSpecificConfig> optionalPreviousConfig =
+          previousConfigDoc.flatMap(this::convertToContextSpecificConfig);
+      long updateTimestamp = clock.millis();
+      long creationTimestamp =
+          previousConfigDoc
+              .filter(configDocument -> !ConfigServiceUtils.isNull(configDocument.getConfig()))
+              .map(ConfigDocument::getCreationTimestamp)
+              .orElse(updateTimestamp);
+      long newVersion =
+          previousConfigDoc
+              .map(ConfigDocument::getConfigVersion)
+              .map(previousVersion -> previousVersion + 1)
+              .orElse(1L);
+      Key latestDocKey = new ConfigDocumentKey(configResourceContext, newVersion);
+      ConfigDocument latestConfigDocument =
           new ConfigDocument(
-              configResource.getResourceName(),
-              configResource.getResourceNamespace(),
-              configResource.getTenantId(),
-              configResource.getContext(),
-              configVersion,
+              configResourceContext.getConfigResource().getResourceName(),
+              configResourceContext.getConfigResource().getResourceNamespace(),
+              configResourceContext.getConfigResource().getTenantId(),
+              configResourceContext.getContext(),
+              newVersion,
               userId,
               latestConfig,
               creationTimestamp,
               updateTimestamp);
-      collection.upsert(key, configDocument);
-      return getInternalContextSpecificConfig(configDocument, existingConfig);
+      collection.upsert(latestDocKey, latestConfigDocument);
+      return optionalPreviousConfig
+          .map(previousConfig -> this.buildUpsertResult(latestConfigDocument, previousConfig))
+          .orElseGet(() -> this.buildUpsertResult(latestConfigDocument));
     }
   }
 
   @Override
-  public ContextSpecificConfig getConfig(ConfigResource configResource) throws IOException {
-    return getConfig(configResource, getLatestVersion(configResource));
+  public List<UpsertedConfig> writeAllConfigs(
+      Map<ConfigResourceContext, Value> resourceContextValueMap, String userId) throws IOException {
+    // TODO keep pushing this down into doc store
+    List<UpsertedConfig> list = new ArrayList<>(resourceContextValueMap.size());
```

Comment:
so would docstore be able to leverage bulk upsert APIs ?

Reply:
That's the idea - just split the PR. The current PR handles all the API changes, which is a fairly significant refactor, but doesn't really change the implementation in any meaningful way. A future PR needs only swap out this implementation to use the doc store's bulk upsert.